### PR TITLE
Discovery of services in Kubernetes OKAPI-1045

### PR DESCRIPTION
**Pull request by adamdickmeiss**
https://issues.folio.org/browse/OKAPI-1045

### CODE REVIEW

File: doc/guide.md [hunk 1]
```diff
@@ -3015,6 +3027,58 @@ systemctl daemon-reload
 systemctl restart docker
 ```
 
+
+### Kubernetes integration
+
+Okapi 4.13.0 has support for service discovery with Kubernetes.
```
Comment (julianladisch) [inline doc/guide.md:3033]:
```suggestion
Since version 4.13.0 Okapi has support for service discovery with Kubernetes.
```

File: doc/guide.md [hunk 2]
```diff
@@ -3015,6 +3027,58 @@ systemctl daemon-reload
...
+
+Okapi uses the Kubernetes service API to get information about endpoints
+of modules. These endpoints will be populated to Okapi discovery sub-system
+in a "refresh" operation. The refresh operation removes and adds relevant
+services. The Kubernetes discovered services have instId `kube_` prefix, so
+that services otherwise manually added to the Okapi are preserved by the
+refresh operation.
+
+The refresh operation happens periodically and at certain important stages,
+such as when before an install operation takes place and when Okapi starts.
```
Comment (julianladisch) [inline doc/guide.md:3047]:
```suggestion
such as before an install operation takes place and when Okapi starts.
```

File: doc/guide.md [hunk 3]
```diff
@@ -3015,6 +3027,58 @@ systemctl daemon-reload
...
+    - name: http
+      port: 8081
+      protocol: TCP
+  selector:
+    app: mod-users
+```
+
+
+The first http based port is used. At this stage if there modules that are
+https based, that's unsupported.
```
Comment (julianladisch) [inline doc/guide.md:3080]:
```suggestion
The first http based port is used. At this stage https based modules are
unsupported.
```

File: okapi-core/src/main/java/org/folio/okapi/managers/KubernetesManager.java [hunk 4]
```diff
@@ -0,0 +1,250 @@
...
+   */
+  public KubernetesManager(DiscoveryManager discoveryManager, JsonObject config) {
+    this.discoveryManager = discoveryManager;
+    refreshInterval = Config.getSysConfInteger(ConfNames.KUBE_REFRESH_INTERVAL, 30000, config);
+    fname = Config.getSysConf(ConfNames.KUBE_CONFIG, null, config);
+    token = Config.getSysConf(ConfNames.KUBE_TOKEN, null, config);
+    server = Config.getSysConf(ConfNames.KUBE_SERVER, null, config);
+    namespace = Config.getSysConf(ConfNames.KUBE_NAMESPACE, "default", config);
+    webClientOptions = new WebClientOptions()
+        .setTrustAll(Config.getSysConfBoolean(ConfNames.HTTP_CLIENT_TRUST_ALL, false, config));
```
Comment (julianladisch) [inline okapi-core/src/main/java/org/folio/okapi/managers/KubernetesManager.java:51]: Never use setTrustAll(true).
This opens a back-door and defeats the security SSL/TLS is used for.

Example how to properly use https: https://github.com/folio-org/okapi/blob/v4.12.0/okapi-core/src/main/java/org/folio/okapi/service/impl/PostgresHandle.java#L70-L80

Supporting https might be split off into a separate pull request: https://issues.folio.org/browse/OKAPI-1071

**Review by julianladisch — CHANGES_REQUESTED** [review]

File: okapi-core/src/main/java/org/folio/okapi/managers/DiscoveryManager.java [hunk 5]
```diff
@@ -150,18 +150,12 @@ public void setModuleManager(ModuleManager mgr) {
    * </p>
    */
   private Future<DeploymentDescriptor> addAndDeploy0(DeploymentDescriptor dd) {
```
[the remark this answers is not in the document]
Comment (adamdickmeiss) [reply okapi-core/src/main/java/org/folio/okapi/managers/DiscoveryManager.java:152]: Yeah. It's a little messy.

File: okapi-core/src/main/java/org/folio/okapi/managers/KubernetesManager.java (hunk 4 above)
Comment (adamdickmeiss) [reply okapi-core/src/main/java/org/folio/okapi/managers/KubernetesManager.java:51]: The Kubernetes API server *is* https based. So we need something we can do for https to work for us in "less secure" invironments such as developments. We can wait with the modules, but why would't that use the same strategy.

File: okapi-core/src/main/java/org/folio/okapi/managers/KubernetesManager.java [hunk 6]
```diff
@@ -0,0 +1,250 @@
...
+  Future<List<DeploymentDescriptor>> getEndpoints() {
+    String uri = server + "/api/v1/namespaces/" + namespace + "/endpoints";
+    HttpRequest<Buffer> abs = webClient.getAbs(uri);
+    if (token != null) {
+      abs.putHeader("Authorization", "Bearer " + token);
+    }
+    return abs.putHeader("Accept", "application/json")
+        .expect(ResponsePredicate.SC_OK)
+        .expect(ResponsePredicate.JSON)
+        .send().map(res -> parseItems(res.bodyAsJsonObject()));
```
[the remark this answers is not in the document]
Comment (adamdickmeiss) [reply okapi-core/src/main/java/org/folio/okapi/managers/KubernetesManager.java:185]: ok?

File: okapi-core/src/main/java/org/folio/okapi/managers/KubernetesManager.java [hunk 7]
```diff
@@ -0,0 +1,250 @@
...
+  }
+
+  Future<Void> readKubeConfig(Vertx vertx, String fname) {
+    if (fname == null) {
+      return Future.succeededFuture();
+    }
+    FileSystem fs = vertx.fileSystem();
+    return fs.readFile(fname).compose(content -> {
+      YamlProcessor yamlProcessor = new YamlProcessor();
+      return yamlProcessor.process(vertx, null, content);
```
[the remark this answers is not in the document]
Comment (adamdickmeiss) [reply okapi-core/src/main/java/org/folio/okapi/managers/KubernetesManager.java:71]: sure

File: okapi-core/src/main/java/org/folio/okapi/managers/KubernetesManager.java [hunk 8]
```diff
@@ -0,0 +1,250 @@
...
+          removeList.add(dd);
+        }
+      }
+    }
+    for (DeploymentDescriptor dd : incoming) {
+      String key = dd.getSrvcId() + "-" + dd.getInstId();
+      if (instances.contains(key)) {
+        addList.add(dd);
+      }
+    }
```
[the remark this answers is not in the document]
Comment (adamdickmeiss) [reply okapi-core/src/main/java/org/folio/okapi/managers/KubernetesManager.java:248]: ok

File: okapi-core/src/main/java/org/folio/okapi/managers/DiscoveryManager.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (adamdickmeiss) [reply okapi-core/src/main/java/org/folio/okapi/managers/DiscoveryManager.java:152]: done

File: okapi-core/src/main/java/org/folio/okapi/managers/KubernetesManager.java (hunk 4 above)
Comment (adamdickmeiss) [reply okapi-core/src/main/java/org/folio/okapi/managers/KubernetesManager.java]: setTrustAll no longer in use. with server PEM for the API server, that gets me going for local microk8s environment.

Comment (julianladisch) [reply okapi-core/src/main/java/org/folio/okapi/managers/KubernetesManager.java]: Nice, thanks!

File: doc/guide.md [hunk 9]
```diff
@@ -2852,6 +2853,18 @@ before 4.10.0.
...
 The value is a boolean - `true` for enable, `false` for disable.  Default is `true`.
+* `kube_config`: Filename/resource which is the Kubernetes configuration
+to use for Kubernetes integration. If defined, Okapi will perform
+service discovery using Kubernetes.
+See [kubernetes](#kubernetes-integration) section.
+* `kube_server_url`: Kubernetes API server URL. If defined, Okapi will perform
+service discovery using Kubernetes.
+see [kubernetes](#kubernetes-integration) section.
+* `kube_server_pem`: PEM file representing server certificate for Kube API server.
+* `kube_token`: Kubernetes API token. If unset, no Autorization will be provided.
```
Comment (julianladisch) [inline doc/guide.md:2864]:
```suggestion
* `kube_token`: Kubernetes API token. If unset, no Authorization will be provided.
```

**Review by julianladisch — APPROVED** [review]
Looks good, thanks!
